### PR TITLE
Validate records locally against schema before submitting ?

### PR DESCRIPTION
**Pull request by Natim**
Currently the scripts assumes that the server will validate the schema.

It should either check that the server has the json validation capability, or validate records locally before submitting them

### CODE REVIEW

File: amo2kinto/tests/test_importer.py [hunk 1]
```diff
@@ -274,6 +280,37 @@ def test_sync_records_calls_the_scenario():
...
+            return_value=[]):
+        with mock.patch('amo2kinto.importer.push_changes'):
+
+            sync_records(RECORDS['addons'],
+                         FIELDS['addons'],
+                         mock.sentinel.kinto_client,
+                         mock.sentinel.bucket,
+                         mock.sentinel.collection,
+                         SCHEMAS['addons'],
+                         mock.sentinel.permissions)
```
Comment (leplatrem) [inline amo2kinto/tests/test_importer.py]: missing an assertion here ?

File: amo2kinto/tests/test_importer.py [hunk 2]
```diff
@@ -274,6 +280,37 @@ def test_sync_records_calls_the_scenario():
...
+
+
+def test_sync_records_validate_records_against_schema_and_fails_if_wrong():
+    with mock.patch(
+            'amo2kinto.importer.get_kinto_records',
+            return_value=[]):
+        with mock.patch('amo2kinto.importer.push_changes'):
+
```
Comment (leplatrem) [inline amo2kinto/tests/test_importer.py]: nit: use `addCleanup(patch.stop)` pattern ?

File: amo2kinto/tests/test_importer.py (hunk 1 above)
Comment (Natim) [inline amo2kinto/tests/test_importer.py]: I can add a try catch to make it expecit that it shouldn't raise a ValidationError.

File: amo2kinto/importer.py [hunk 3]
```diff
@@ -28,6 +29,11 @@ def sync_records(amo_records, fields,
                  kinto_client, bucket, collection, schema, permissions):
 
     amo_records = prepare_amo_records(amo_records, fields)
+
+    if schema:
+        for record in amo_records:
+            jsonschema.validate(record, schema)
```
Comment (leplatrem) [inline amo2kinto/importer.py]: don't we want to validate the `to_create` list only ?

Comment (Natim) [inline amo2kinto/importer.py]: That's the question I was asking myself, in anycase it shouldn't fails or it means either the AMO database is broken (which is the case on the dev server for instance) either the schema is not up-to-date.

To answer this question we should answer: Why do we want local validation?

- Is it to make sure the kinto database only contains valid records?
- Is it to make sure records we read from AMO are valids?

Comment (leplatrem) [inline amo2kinto/importer.py]: > To answer this question we should answer: Why do we want local validation?

Yes, it is a very useful safety check. As you said, either to check the existing records against a new schema, or check that the new ones are compatibles with the schema in place.

File: amo2kinto/tests/test_importer.py (hunk 1 above)
Comment (leplatrem) [inline amo2kinto/tests/test_importer.py]: A comment can be enough then

File: amo2kinto/importer.py (hunk 3 above)
Comment (Natim) [inline amo2kinto/importer.py]: Does it means I should leave it like that? Or that I should validate to_create records only?

**Comment by leplatrem** [comment]
r+

**Pull request merged**